src/static_array.cr

```diff
@@ -160,13 +160,50 @@ struct StaticArray(T, N)
     N
   end
 
+  # Replaces every element in `self` with the given *value*. Returns `self`.
+  #
+  # ```
+  # array = StaticArray(Int32, 3).new 0 # => StaticArray[0, 0, 0]
+  # array.fill(2)                       # => StaticArray[2, 2, 2]
+  # array                               # => StaticArray[2, 2, 2]
+  # ```
+  def fill(value : T)
+    {% if Int::Primitive.union_types.includes?(T) || Float::Primitive.union_types.includes?(T) %}
+      if value == 0
+        to_unsafe.clear(size)
+
+        self
+      else
+        fill { value }
+      end
+    {% else %}
+      fill { value }
+    {% end %}
```

Comment:
I'm not sure this optimization for numerical primitives is really worth it, but the code could at least be more concise by not repeating the delegation.
```suggestion
    {% if Int::Primitive.union_types.includes?(T) || Float::Primitive.union_types.includes?(T) %}
      if value == 0
        to_unsafe.clear(size)

        return self
      end
    {% end %}
      fill { value }
```